zcash_client_backend/src/scanning.rs

```diff
@@ -365,22 +365,57 @@ pub(crate) fn scan_block_with_runner<
             if m.sapling_commitment_tree_size == 0 {
                 None
             } else {
-                let block_note_count: u32 = block
+                let block_output_count: u32 = block
                     .vtx
                     .iter()
-                    .map(|tx| {
-                        u32::try_from(tx.outputs.len()).expect("output count cannot exceed a u32")
-                    })
-                    .sum();
-                Some(m.sapling_commitment_tree_size - block_note_count)
+                    .map(|tx| tx.outputs.len())
+                    .sum::<usize>()
+                    .try_into()
+                    .expect("Sapling output count cannot exceed a u32");
+                Some(m.sapling_commitment_tree_size - block_output_count)
             }
         })
         .or_else(|| prior_block_metadata.map(|m| m.sapling_tree_size()))
+        .or_else(|| {
+            // If we're below Sapling activation, or Sapling activation is not set, the tree size is zero
+            params
+                .activation_height(NetworkUpgrade::Sapling)
+                .map_or(Some(0), |h| if h > cur_height { Some(0) } else { None })
+        })
         .ok_or(ScanError::TreeSizeUnknown {
             protocol: ShieldedProtocol::Sapling,
             at_height: cur_height,
         })?;
 
+    let mut orchard_commitment_tree_size = block
+        .chain_metadata
+        .as_ref()
+        .and_then(|m| {
+            if m.orchard_commitment_tree_size == 0 {
+                None
```

Comment:
Same comment as above; should it be this?
```suggestion
                Some(0)
```